baseplate/clients/redis.py

```diff
@@ -269,7 +269,39 @@ def __init__(
     # pylint: disable=arguments-differ
     def execute(self, **kwargs: Any) -> Any:
         with self.server_span.make_child(self.trace_name):
-            return super().execute(**kwargs)
+
+            success = "true"
+            start_time = perf_counter()
+            labels = {
+                f"{PROM_LABELS_PREFIX}_command": "pipeline",
+                f"{PROM_LABELS_PREFIX}_client_name": self.connection_pool.connection_kwargs.get(
+                    "client_name", ""
+                ),
+                f"{PROM_LABELS_PREFIX}_database": self.connection_pool.connection_kwargs.get(
+                    "db", ""
+                ),
+                f"{PROM_LABELS_PREFIX}_type": "standalone",
+            }
+            num_reqs = len(self.command_stack)
+            commands = [args[0] for (args, options) in self.command_stack]
+            ACTIVE_REQUESTS.labels(**labels).inc()
+            try:
+                return super().execute(**kwargs)
+            except:  # noqa: E722
+                success = "false"
+                raise
+            finally:
+                ACTIVE_REQUESTS.labels(**labels).dec()
+                for command in commands:
+                    result_labels = {
+                        **labels,
+                        f"{PROM_LABELS_PREFIX}_success": success,
+                        f"{PROM_LABELS_PREFIX}_command": command,
+                    }
+                    REQUESTS_TOTAL.labels(**result_labels).inc()
```

Comment:
Why are we doing this per command in the pipeline? Wouldn't the other commands be tracked separately from the pipeline itself via the `super` call?

Reply:
> Why are we doing this per command in the pipeline? Wouldn't the other commands be tracked separately from the pipeline itself via the `super` call?

If you look at what that `super` call does it never hits any code path we've already instrumented. We assumed it did, which is why we didn't instrument this to begin with. What redis actually does is [pack the commands into one request](https://github.com/redis/redis-py/blob/f9f9d06c9951f8536bf9321dcebc96759eae03e0/redis/asyncio/client.py#L1261-L1266).

Hence its one "active request"/connection, but several commands being executed. See my other comment about including the individual names of commands executed, it's my preference to remove it but that's not compliant with the spec.

Reply:
Actually, looking at the clustered Redis implementation (only), on retry it [can call `execute_command` on `RedisCluster`](https://github.com/redis/redis-py/blob/fd9fea6bc07bf0970a5a42c5ec1788272446910c/redis/cluster.py#L1998-L2026). As I understand it we want this to count in the metrics for every attempt, but potentially there are pitfalls.

Reply:
Cool, thanks for the explanation!

Reply:
Ideally, it doesn't make sense to associate the time the entire pipeline with each individual command in the pipeline. If folks want instrumentation per command they can provide a connection with instrumentation. Do you know how this work currently for Wavefront users?